### DISCUSSION TITLE
Batch transcribe mp3 files

### DISCUSSION THREAD

**Discussion by dvis003** [openai/whisper#256] [Q&A]
Hi All, 
I came across [this reddit thread](https://www.reddit.com/r/MachineLearning/comments/xl7mfy), where author benchmarks whisper against peers and implements batching. [In the discussion](https://www.reddit.com/r/MachineLearning/comments/xl7mfy/comment/iqf34zo), which follows he showed his implementation:

```
def process_audio(audio_list):
mels = []
for audio in audio_list:
    audio = whisper.load_audio(audio)
    audio = whisper.pad_or_trim(audio)
    mel = whisper.log_mel_spectrogram(audio).unsqueeze(0)
    mels.append(mel)
mels = torch.cat(mels, dim=0)
return mels


with torch.no_grad():
    mels = process_audio(batch).to(device)
    results = whisper.decode(model, mels, options)
    print([result.text for result in results])
```
However, I struggle to input correct dimension to the respective functions
```
unsqueeze(0)
torch.cat(mels, dim=0)
```
for the batch size of 8 mp3 files. I am complete noob in the ML space, hence would appreciate if somebody can donate their time to write out a working implementation for (I am fine with hardcoded 8 files in the batch) the batch processor.

Fine paying some 🍺 money for it, but also believe this would benefit more causal whisper community.

Cheers!

**Comment by fortypercnt** [comment]
whisper.pad_or_trim(audio) trims any audio beyond the first 30s. Unless all files in batch are shorter than 30s, I think this code does not do what it's supposed to. As for dimensions the code looks correct. It would help if you could share the error messages you are getting.

The code works fine if you preprocess files in batch into 30s segments, but the translation / transcription will be disjointed between segments. A better way is to use model.transcribe(...) which takes care internally of splitting longer audio into segments and conditioning the model on previous segments. I don't see an easy way of enabling batch processing with this though, since it introduces dependency between segments.

**Reply by dvis003** [reply]
Thank you for the explanation, it makes way more sense now on why it failed.
Actually for me the end goal is to speed up the transcription workflow. I just have a simple idea in mind of batch processing a bunch of podcasts that a listen to, for note taking and future reference. In this light 
>  transcription will be disjointed between segments
is kinda deal breaker and thank you for pointing this out...
My ideas is as simple as, take a tuple of say 8 mp3 files (each of varying length) and solve for the transcription speed. 
Since per my measurements it takes ca. 3min to transcribe a single 90min episode on a rented single T4 or A10, but I have read that most of it is initially due model load time, naturally leads the more episodes staffed in, the cheaper the cost per single episode gets.
That's all I am looking for code wise. Just load a bunch of mp3's as an input and spit out a bunch of txt files as an output in the least time possible (tiny.en model in my tests is more than good enough for the task)

**Reply by jianfch** [reply]
alternative solution to speeding up transcription with multiple audio files, discussed in #153

**Comment by Blair-Johnson** [comment]
I've made a fork that offers a batch-parallel implementation of the `model.transcribe()` method. You can check out the discussion at https://github.com/openai/whisper/discussions/662.

**Comment by eduardofc** [comment]
If it helps... 

I recommend you to install datasets:

```
from datasets import Dataset, Audio

files = ... (a list of mp3-paths)
dataset = Dataset.from_dict({"audio": files}).cast_column("audio", Audio(sampling_rate=16000))
```

and transformers. Specially, I recommend you the use of its pipelines:

```
from tqdm import tqdm
from transformers import pipeline
from transformers.pipelines.pt_utils import KeyDataset

pipe = pipeline( 
  task="automatic-speech-recognition", 
  model="openai/whisper-large", 
  # chunk_length_s=30, 
  # device=device, 
  # ignore_warning=True 
) 

transcriptions = []
with torch.no_grad():
    for tt in tqdm(pipe(KeyDataset(dataset, "audio"), batch_size=10, truncation=True)):
        transcriptions.append(tt['text'])
```
